### PR TITLE
Security Group support for OpenStack Load Balancers

### LINKED ISSUE #29745: OpenStack LBaaS integration missing Security Group Configuration
Currently in LBaaS in openstack (v1 and v2), it puts the LBaaS port in the default security group only. Most environments have that set not to allow any traffic. There's a bug open to ask for a security group when creating an LB, but it hasn't gone anywhere. This means if you use this plugin with kubernetes, it with create the LB and assign a floating ip, but no one will be able to access it. the best fix is allow the config file to specify a security group id- the plugin would get the port_id of the LB after its created and then use update-port (http://gophercloud.io/docs/networking/#update-port) to add the security group to the port. if no security group was supplied in the config file it would behave as it does now.

### PR DESCRIPTION
**Pull request by grahamhayes**
**Add Security Group Support for OpenStack Load Balancers**:

fixes #29745
adds OpenStack support to the work done in #20392

**Release note**:

```
This allows security groups to be created and attached to the neutron
port that the load balancer is using on the subnet.

The security group ID that is assigned to the nodes needs to be
provided, to allow for traffic from the load balancer to the nodePort
to be reflected in the rules.

This adds two config items to the LoadBalancer options -

ManageSecurityGroups (bool)
NodeSecurityGroupID  (string)
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31921)

### CODE REVIEW

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 1]
```diff
@@ -240,6 +243,37 @@ func getLoadbalancerByName(client *gophercloud.ServiceClient, name string) (*loa
...
+
+	var securityRules []rules.SecGroupRule
+
+	err := pager.EachPage(func(page pagination.Page) (bool, error) {
+		ruleList, err := rules.ExtractRules(page)
+		if err != nil {
+			return false, err
+		}
+
+		for _, rule := range ruleList {
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Extremely minor point, but you can append lists with:
    securityRules = append(securityRules, ruleList...)

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 2]
```diff
@@ -475,6 +530,86 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *api.Ser
...
+					EtherType:      "IPv4",
+				}
+
+				_, err := rules.Create(lbaas.network, lbSecGroupRuleCreateOpts).Extract()
+
+				if err != nil {
+					return nil, err
+				}
+
+				lbaas.createNodeSecurityGroup(int(port.Port), string(port.Protocol), lbSecGroup.ID)
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: I think this is effectively a duplicate of the above `rules.Create()`.  Drop this one, and the `createNodeSecurityGroup` function above?

Comment (grahamhayes) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: What this does is create the rules for traffic from the loadbalancer to the node.

What the `rules.Create()` does is allow traffic from outside to the loadbalancer.

There has to be 2 sets of rules, as the ports for each side are different.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 3]
```diff
@@ -475,6 +530,86 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *api.Ser
...
+
+			for _, sourceRange := range sourceRanges.StringSlice() {
+				lbSecGroupRuleCreateOpts := rules.CreateOpts{
+					Direction:      "ingress",
+					PortRangeMax:   int(port.Port),
+					PortRangeMin:   int(port.Port),
+					Protocol:       strings.ToLower(string(port.Protocol)),
+					RemoteIPPrefix: string(sourceRange),
+					SecGroupID:     lbSecGroup.ID,
+					EtherType:      "IPv4",
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: This of course assumes `sourceRange` is IPv4.  I suggest:
    ethertype := "IPv4"
    if sourceRange.IP.To4() == nil {
        ethertype = "IPv6"
    }

That's not great either, since we make no effort to check the address family of floating IP and loadbalancer VIP above, but it's a start..

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 4]
```diff
@@ -475,6 +530,86 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *api.Ser
...
+			return nil, err
+		}
+
+		update_opts := neutron_ports.UpdateOpts{SecurityGroups: append(port.SecurityGroups, lbSecGroup.ID)}
+
+		res := neutron_ports.Update(lbaas.network, port.ID, update_opts)
+
+		if res.Err != nil {
+			glog.Errorf("Error occured updating port: %s", portID)
+			return nil, err
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Wrong error, should be:

```
return nil, res.Err
```

Alternatively, elsewhere in the openstack plugin we've used something more like this:
```
_, err := neutron_ports.Update(...)
if err != nil {
   glog.Errorf("Error ...")
   return nil, err
}
```

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 5]
```diff
@@ -475,6 +530,86 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *api.Ser
...
+
+		lbSecGroupCreateOpts := groups.CreateOpts{
+			Name:        getSecurityGroupName(clusterName, apiService),
+			Description: fmt.Sprintf("Securty Group for %v Service LoadBalancer", apiService.Name),
+		}
+
+		lbSecGroup, err := groups.Create(lbaas.network, lbSecGroupCreateOpts).Extract()
+
+		if err != nil {
+			return nil, err
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: This, and all other early returns below should unwind and cleanup the loadbalancer.
```go
_ = lbaas.EnsureLoadBalancerDeleted(clusterName, apiService)
return nil, err
```

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 6]
```diff
@@ -591,6 +741,11 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
 			if err != nil {
 				return err
 			}
+			if lbaas.opts.ManageSecurityGroups {
+				// Allow traffic from the
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Unfinished comment?

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 7]
```diff
@@ -591,6 +741,11 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
...
+				lbaas.createNodeSecurityGroup(int(port.Port), string(port.Protocol), lbSecGroup.ID)
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: error return?

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 8]
```diff
@@ -591,6 +741,11 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
 			if err != nil {
 				return err
 			}
+			if lbaas.opts.ManageSecurityGroups {
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Why is this here?  Won't the security group already have been created in `EnsureLoadBalancer`?

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 9]
```diff
@@ -604,6 +759,26 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
...
+			// Delete the rules in the Node Security Group
+			opts := rules.ListOpts{
+				SecGroupID:    lbaas.opts.NodeSecurityGroupID,
+				RemoteGroupID: lbSecGroup.ID,
+				PortRangeMax:  int(port.Port),
+				Protocol:      string(port.Protocol),
+			}
+			rules_, err := getSecurityGroupRules(lbaas.network, opts)
+
+			if err != nil {
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: "not found" should not be treated as an error here - it just means we've already partially deleted the loadbalancer, failed, and being retried.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 10]
```diff
@@ -604,6 +759,26 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
...
+				glog.Errorf("Error looking up security group rules for loadbancer IaaS ID: %s", loadbalancer.ID)
+				return err
+			}
+
+			for _, rule := range rules_ {
+				res := rules.Delete(lbaas.network, rule.ID)
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Ditto, "not found" should not be treated as an error here.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 11]
```diff
@@ -759,6 +934,44 @@ func (lbaas *LbaasV2) EnsureLoadBalancerDeleted(clusterName string, service *api
...
+
+	// Delete the Security Group
+	if lbaas.opts.ManageSecurityGroups {
+		// Generate Name
+		lbSecGroupName := getSecurityGroupName(clusterName, service)
+		lbSecGroupID, err := groups.IDFromName(lbaas.network, lbSecGroupName)
+
+		if err != nil {
+			// We were getting this group to delete it anyway - let just continue
+			glog.Errorf("Error occured finding securty group: %s", lbSecGroupName)
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: typo: "securty"
Also: include `err` in the error message, otherwise we're losing valuable debugging information.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 12]
```diff
@@ -604,6 +759,26 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
 			if err != nil && !isNotFound(err) {
 				return err
 			}
+			// Delete the rules in the Node Security Group
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Why is this here?
We're removing a backend host from the LB pool, but I don't see anything node-specific about the secgroup rule you're deleting...

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 13]
```diff
@@ -759,6 +934,44 @@ func (lbaas *LbaasV2) EnsureLoadBalancerDeleted(clusterName string, service *api
...
+		lbSecGroupName := getSecurityGroupName(clusterName, service)
+		lbSecGroupID, err := groups.IDFromName(lbaas.network, lbSecGroupName)
+
+		if err != nil {
+			// We were getting this group to delete it anyway - let just continue
+			glog.Errorf("Error occured finding securty group: %s", lbSecGroupName)
+
+		} else {
+			lbSecGroup := groups.Delete(lbaas.network, lbSecGroupID)
+			if lbSecGroup.Err != nil {
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: "not found" should not be treated as an error here.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 14]
```diff
@@ -475,6 +530,86 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *api.Ser
...
+
+				lbaas.createNodeSecurityGroup(int(port.Port), string(port.Protocol), lbSecGroup.ID)
+			}
+
+		}
+
+		var port neutron_ports.Port
+
+		// Get the port ID
+		portID, err := getPortIDByIP(lbaas.network, loadbalancer.VipAddress)
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Rather than lookup the port, get the ID, then look up the port again (below) - I suggest adapting `getPortIDByIP` to return the full port object (and rename the function appropriately).

**Comment by anguslees** [comment]
@grahamhayes: Nice change, I look forward to getting this in.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 15]
```diff
@@ -604,6 +759,26 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
 			if err != nil && !isNotFound(err) {
 				return err
 			}
+			// Delete the rules in the Node Security Group
+			opts := rules.ListOpts{
+				SecGroupID:    lbaas.opts.NodeSecurityGroupID,
+				RemoteGroupID: lbSecGroup.ID,
+				PortRangeMax:  int(port.Port),
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: I think this should be `port.NodePort`

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 2 above)
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Oh of course, I missed `RemoteGroupID` - thanks.
In this case, I think it should be `port.NodePort`

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 16]
```diff
@@ -282,6 +316,25 @@ func waitLoadbalancerDeleted(client *gophercloud.ServiceClient, loadbalancerID s
...
 
+func (lbaas *LbaasV2) createNodeSecurityGroup(port int, protocol string, lbSecGroup string) error {
+	nodeSecGroupRuleCreateOpts := rules.CreateOpts{
+		Direction:     "ingress",
+		PortRangeMax:  port,
+		PortRangeMin:  port,
+		Protocol:      strings.ToLower(protocol),
+		RemoteGroupID: lbSecGroup,
+		SecGroupID:    lbaas.opts.NodeSecurityGroupID,
+		EtherType:     "IPv4",
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Please create a duplicate rule to allow IPv6 too.  We make no effort to filter to only IPv4 addresses in `NodeAddresses` and it's quite possible that traffic goes IPv4 to the LB and then IPv6 to the backend node.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 12 above)
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: (Changed my mind and I'm restoring a deleted comment)
Why is this here?  There's nothing node-specific about the rules we're about to delete.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 17]
```diff
@@ -282,6 +316,25 @@ func waitLoadbalancerDeleted(client *gophercloud.ServiceClient, loadbalancerID s
 	}
 }
 
+func (lbaas *LbaasV2) createNodeSecurityGroup(port int, protocol string, lbSecGroup string) error {
```
Comment (dagnello) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: this can be a helper function instead of an instance function of LbaasV2

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 1 above)
Comment (grahamhayes) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 2 above)
Comment (grahamhayes) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Yup - in my cleanup I seem to have moved it to the wrong section - done.

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 3 above)
Comment (grahamhayes) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 4 above)
Comment (grahamhayes) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 5 above)
Comment (grahamhayes) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 8 above)
Comment (grahamhayes) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Yeap. Removed

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 9 above)
Comment (grahamhayes) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 10 above)
Comment (grahamhayes) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Ditto - Done

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 11 above)
Comment (grahamhayes) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 13 above)
Comment (grahamhayes) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go (hunk 14 above)
Comment (grahamhayes) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Done

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 18]
```diff
@@ -475,6 +530,86 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *api.Ser
...
 	}
 
+	if lbaas.opts.ManageSecurityGroups {
+
+		lbSecGroupCreateOpts := groups.CreateOpts{
+			Name:        getSecurityGroupName(clusterName, apiService),
+			Description: fmt.Sprintf("Securty Group for %v Service LoadBalancer", apiService.Name),
+		}
+
+		lbSecGroup, err := groups.Create(lbaas.network, lbSecGroupCreateOpts).Extract()
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: Somewhere we need to add a rule to this secgroup to allow (at least) ICMP frag-needed (and perhaps equivalent for IPv6), otherwise PMTUD will be broken.

**Comment by anguslees** [comment]
@grahamhayes: Afaik, port changes turn into a delete+recreate.  My understanding is that UpdateLoadBalancer is only used for pool membership changes.

**Comment by anguslees** [comment]
Review status: 0 of 9 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go, line 595 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31921#-KR6NNu7cYRer9fiudmu:-KR6NNu7cYRer9fiudmv:b-dmax23) ([raw file](https://github.com/kubernetes/kubernetes/blob/9547b9a6153ea55bd8c743c5f1e0f2c17af47076/pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go#L595)):*
> ```Go
> 			}
> 
> 			createNodeSecurityGroup(lbaas.network, lbaas.opts.NodeSecurityGroupID, int(port.NodePort), string(port.Protocol), lbSecGroup.ID)
> ```

check error return

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31921)*

**Comment by anguslees** [comment]
Review status: 0 of 9 files reviewed at latest revision, 14 unresolved discussions.

---

*[pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go, line 104 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31921#-KR6QidpjZCjxQPLQ1Fl:-KR6QidqrgSU_zNtwu6w:b4xyyjo) ([raw file](https://github.com/kubernetes/kubernetes/blob/9547b9a6153ea55bd8c743c5f1e0f2c17af47076/pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go#L104)):*
> ```Go
> 	})
> 
> 	return targetPort, err
> ```

```go
if err == nil && targetPort == nil {
   err = ErrNotFound
}
```
... otherwise we have a nil port object that no-one is going to check for, leading to runtime crashes :(

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31921)*

**Comment by anguslees** [comment]
Review status: 0 of 9 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go, line 573 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31921#-KR129i0RjU-crRqVmZJ-r2-573:-KR6Rk2X4412DkUAE-ST:bqivl7w) ([raw file](https://github.com/kubernetes/kubernetes/blob/19fcb9160867717cd8e615ba94e9e4eb60e4b955/pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go#L573)):*
<details><summary><i>Previously, grahamhayes (Graham Hayes) wrote…</i></summary>
> Done
</details>
(still need to adapt this part to use the new `getPortByIP`)

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31921)*

File: pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go [hunk 19]
```diff
@@ -97,7 +101,15 @@ func getPortIDByIP(client *gophercloud.ServiceClient, ipAddress string) (string,
 		return true, nil
 	})
```
Comment (anguslees) [inline pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go]: (copying comment here from Reviewable weird app output.  I won't use that to add line comments in future :/ )

```go
if err == nil && targetPort == nil {
   err = ErrNotFound
}
```
... otherwise we have a nil port object that no-one is going to check for, leading to runtime crashes :(

**Comment by anguslees** [comment]
Just https://github.com/kubernetes/kubernetes/pull/31921/files#r78289021, and then I think we're good to go.  Thanks for your patience...

**Comment by grahamhayes** [comment]
Great. I will fix that up.

Should I squash the commits?

**Comment by anguslees** [comment]
@grahamhayes: yes please.  Keep the vendor/ update separate, but squash the review updates.

**Comment by grahamhayes** [comment]
@anguslees should all be updated, and I rebased latest master in.

**Comment by anguslees** [comment]
lgtm

**Comment by grahamhayes** [comment]
@anguslees I think it needs an "ok to test" as well

**Comment by lavalamp** [comment]
ok to test

**Comment by anguslees** [comment]
@grahamhayes: You need to update the licenses file as part of adding to `vendor/`.  See the `./hack/verify-godep-licenses.sh` failure for the required update command.

**Comment by grahamhayes** [comment]
@k8s-bot gci gce e2e test this

Looks like #33627

**Comment by grahamhayes** [comment]
@k8s-bot gke e2e test this
Looks like #33617

**Comment by anguslees** [comment]
@k8s-bot gke e2e test this

**Comment by grahamhayes** [comment]
@k8s-bot gci gce e2e test this

**Comment by anguslees** [comment]
@k8s-bot gci gce e2e test this

**Comment by dagnello** [comment]
@anguslees this pr has been rebased and is ready to merge.  thanks

**Comment by chrislovecnm** [comment]
@k8s-bot gci gce e2e test this
